baselines/feddebug/README.md

```diff
@@ -0,0 +1,221 @@
+```
+Following is the graph `non_iid-resnet18-mnist.png` generated by the code:
+
+![FedDebug Malicious Client Localization Non-IID-ResNet18-CIFAR10](figures/non_iid-resnet18-mnist.png)
+
+
+### C. Threshold Impact on Localization
+You can also test the impact of the neuron activation threshold on localization accuracy. A higher threshold decreases the localization accuracy. Total Time Taken: 84 seconds. 
+
+```bash
+python -m feddebug.main device=cuda model=resnet18 feddebug.na_t=0.7
+```
+
+```log
+...
+[2024-10-24 12:21:26,923][flwr][INFO] - ***FedDebug Output Round 2 ***
+[2024-10-24 12:21:26,923][flwr][INFO] - True Malicious Clients (Ground Truth) = ['0']
+[2024-10-24 12:21:26,923][flwr][INFO] - Total Random Inputs = 10
+[2024-10-24 12:21:26,923][flwr][INFO] - Predicted Malicious Clients = {'5': 0.7, '0': 0.3}
+[2024-10-24 12:21:26,923][flwr][INFO] - FedDebug Localization Accuracy = 30.0
+[2024-10-24 12:21:27,773][flwr][INFO] - fit progress: (2, 0.001345307207107544, {'accuracy': 0.9497, 'loss': 0.001345307207107544, 'round': 2}, 31.669926984992344)
+[2024-10-24 12:21:27,773][flwr][INFO] - configure_evaluate: no clients selected, skipping evaluation
+[2024-10-24 12:21:27,773][flwr][INFO] - 
+```
+
+Following is the graph `iid-resnet18-mnist.png` generated by the code:
+![FedDebug Malicious Client Localization IID-ResNet18-MNIST](figures/iid-resnet18-mnist.png)
+
+
+> [!WARNING]
+> FedDebug generates random inputs to localize malicious client(s). Thus, results might vary slightly on each run due to randomness.
+
+
+
+## Limitations and Discussion
+Compared to the current baseline, FedDebug was originally evaluated using only a single round of training. It was not initially tested with Dirichlet partitioning for data distribution, which means it may deliver suboptimal performance under different data distribution settings. Enhancing FedDebug's performance could be achieved by generating more effective random inputs, for example, through the use of Generative Adversarial Networks (GANs).
+
+
+## Application of FedDebug  
+We used FedDebug to detect `backdoor attacks` in Federated Learning, resulting in [FedDefender](https://dl.acm.org/doi/10.1145/3617574.3617858). The code is implemented using the Flower Framework in [this repository](https://github.com/warisgill/FedDefender). We plan to adapt FedDefender to Flower baseline guidelines soon.
+
+## Citation
+
+If you publish work that uses FedDebug, please cite FedDebug as follows:
+
+```bibtex
+@inproceedings{gill2023feddebug,
+  title={{Feddebug: Systematic Debugging for Federated Learning Applications}},
+  author={Gill, Waris and Anwar, Ali and Gulzar, Muhammad Ali},
+  booktitle={2023 IEEE/ACM 45th International Conference on Software Engineering (ICSE)},
+  pages={512--523},
+  year={2023},
+  organization={IEEE}
+}
+```
+
+
+I am very grateful to [Javier Fernandez](https://www.linkedin.com/in/jafermarq/) for the guidance and support in creating this baseline.
```

Comment:
Very kind of you!! but we could remove this :)

Reply:
updated